man/vzctl.8
.\" $Id$
.TH vzctl 8 "May 2011" "@PRODUCT_NAME_SHORT@"
.SH NAME
vzctl \- utility to control a Container.
.SH SYNOPSIS
vzctl [\fB--quiet\fR | \fB--verbose\fR] \fBcreate\fR \fIctid\fR | \fB--ostemplate\fR \fIname\fR}
[\fB--private\fR \fIpath\fR] [\fB--root\fR \fIpath\fR] [\fB--ipadd\fR \fIaddr\fR[\fB/\fImask\fR]] [\fB--hostname\fR \fIname\fR] [\fB--skip_app_templates\fR] [\fB--name\fR \fIname\fR] [\fB--description\fR \fIdesc\fR] [\fB--force\fR]
.PP
vzctl [\fB--quiet\fR | \fB--verbose\fR] \fBreinstall\fR \fIctid\fR [\fB--skipbackup\fR] [\fB--resetpwdb\fR] [\fB--skipscripts\fR]
.PP
vzctl \fBpause\fR \fIctid\fR
.PP
vzctl \fBrunscript\fR \fIctid\fR \fIscript\fR
.PP
vzctl \fBexecaction\fR \fIctid\fR \fIaction\fR
.TP
vzctl [\fB--quiet\fR | \fB--verbose\fR] \fBset\fR \fIctid\fR [\fB--save\fR]
.\" UBC parameters
[\fB-p\fR, \fB--numproc\fR \fIitems\fR[:\fIitems\fR]]
[\fB--vmguarpages\fR \fIpages\fR[:\fIpages\fR]]
[\fB-l\fR, \fB--lockedpages\fR \fIpages\fR[:\fIpages\fR]]
[\fB--privvmpages\fR \fIpages\fR[:\fIpages\fR]]
[\fB--shmpages\fR \fIpages\fR[:\fIpages\fR]]
[\fB-n\fR, \fB--numfile\fR \fIitems\fR[:\fIitems\fR]]
[\fB-f\fR, \fB--numflock\fR \fIitems\fR[:\fIitems\fR]]
[\fB-t\fR, \fB--numpty\fR \fIitems\fR[:\fIitems\fR]]
[\fB-i\fR, \fB--numsiginfo\fR \fIitems\fR[:\fIitems\fR]]
[\fB-e\fR, \fB--numiptent\fR \fInum\fR[:\fInum\fR]]
[\fB--swappages\fR \fInum\fR[:\fInum\fR]]
[\fB--physpages\fR \fIpages\fR[:\fIpages\fR]]
[\fB--ram\fR \fIbytes\fR]
[\fB--swap\fR \fIbytes\fR]
[\fB--numnetif\fR \fInum\fR]
[\fB--memguarantee\fR \fIauto|value\fR]

[\fB--vm_overcommit \fR \fIN\fR]
.br
.\" Disk quota parameters
[\fB--diskspace\fR \fInum\fR]
[\fB--quotaugidlimit\fR \fInum\fR]
[\fB--jquota\fR \fBon\fR|\fBoff\fR]

.br
.\" CPU fairsched parameters
[\fB--cpuunits\fR \fIvalue\fR]
[\fB--cpulimit\fR \fIvalue\fR]
[\fB--cpus\fR \fIvalue\fR]
[\fB--cpumask\fR {\fIN[,N][,N-N]\fR|\fBall\fR|\fBauto\fR}]
[\fB--nodemask\fR {\fIN[,N][,N-N]\fR|\fBall\fR}]
.br
.\" change some files inside CT
[\fB--ipadd\fR \fIaddr\fR[\fB/\fImask\fR]] [\fB--ipdel\fR \fIaddr\fR|\fBall\fR]
.br
[\fB--netif_add\fR <\fIifname\fR[\fI,mac,host_ifname,host_mac\fR]]>]
[\fB--netif_del\fR <\fIifname\fR>]
.br
[\fB--ifname\fR <\fIifname\fR> [\fB--mac\fR <\fImac\fR>] [\fB--host_ifname\fR <\fIname\fR>]
[\fB--host_mac\fR <\fImac\fR>] [\fB--ipadd\fR <\fIaddr\fR>] [\fB--ipdel\fR <\fIaddr\fR>]
[\fB--gw\fR <\fIaddr\fR>] [\fB--gw6\fR <\fIaddr\fR>] [\fB--dhcp\fR <\fByes\fR|\fBno\fR>]
[\fB--dhcp6 <\fByes\fR|\fBno\fR>] [\fB--network\fR <\fIid\fR>]
[\fB--configure\fR <\fInone\fR|\fBall\fR>]]
.br
[\fB--ext_ipadd\fR \fIaddr\fR] [\fB--ext_ipdel\fR \fIaddr\fR|\fBall\fR]
[\fB--hostname\fR \fIname\fR]
[\fB--nameserver\fR \fIaddr\fR]
[\fB--searchdomain\fR \fIname\fR]
[\fB--userpasswd\fR \fIuser\fR:\fIpassword\fR [\fB--crypted\fR]]
.br
.\" onboot
[\fB--onboot\fR \fByes\fR|\fBno\fR]
[\fB--bootorder\fR \fInumber\fR]
.br
.\" traffic shaping accounting
[\fB--rate\fR \fIdev\fR:\fIclass\fR:\fIKbits\fR]
[\fB--ratebound\fR \fByes\fR|\fBno\fR]
.br
.\" mount option
[\fB--root\fR \fIpath\fR]
[\fB--noatime\fR \fByes\fR|\fBno\fR]
[\fB--bindmount_add\fR [\fIsrc\fR:]\fIdst\fR[,\fBnosuid\fR,\fBnoexec\fR,\fBnodev\fR]
[\fB--bindmount_del\fR \fIdst\fR|\fBall\fR]
[\fB--mount_opts\fR \fIopt\fR[,\fIopt\fR]]
.br
.\" devices
[\fB--devnodes\fR \fIdevice\fR:\fBr\fR|\fBw\fR|\fBrw\fR|\fBnone\fR]
[\fB--netdev_add\fR \fIname\fR]
[\fB--netdev_del\fR \fIname\fR]
[\fB--pci_add\fR \fI[domain:]bus:slot.func\fR]
[\fB--pci_del\fR \fIn[domain:]bus:slot.func\fR]
[\fB--disabled\fR \fByes\fR|\fBno\fR]
[\fB--applyconfig\fR \fIname\fR]
[\fB--setmode\fR \fBrestart\fR|\fBignore\fR]
[\fB--description\fR \fIdesc\fR]
[\fB--name\fR \fIname\fR] [\fB--ioprio\fR \fIprio\fR]
[\fB--iolimit\fR \fIlimit\fR] [\fB--iopslimit\fR \fIlimit\fR]
[\fB--features\fR \fIname\fR:\fBon\fR|\fBoff\fR[\fB,\fR...]]
[\fB--distribution\fR \fIdist\fR]
.br
.\" high availability cluster
[\fB--ha_enable\fR \fByes\fR|\fBno\fR]
[\fB--ha_prio\fR \fInumber\fR]
.PP
vzctl [\fB--quiet\fR | \fB--verbose\fR]
\fBdestroy\fR | \fBmount\fR | \fBumount\fR |
\fBstart\fR | \fBrestart\fR |
\fBstatus\fR | \fBenter\fR | \fBconsole\fR |
.PP
vzctl \fBstop\fR \fIctid\fR|\fIname\fR [\fB--fast\fR] [\fB--skip-umount\fR]
.PP
vzctl \fBstart\fR \fIctid\fR|\fIname\fR [\fB--wait\fR] [\fB--skip-fsck\fR]
.PP
vzctl \fBmonitor\fR \fIctid|\fIname\fR|\fB0\fR
.PP
vzctl \fBsuspend\fR \fIctid\fR|\fIname\fR
.PP
vzctl \fBresume\fR \fIctid\fR|\fIname\fR [\fB--skip-fsck\fR]
.PP
vzctl \fBsnapshot\fR \fIctid\fR|\fIname\fR
[\fB--id\fR \fIuuid\fR]
[\fB--name\fR \fIname\fR]
[\fB--description\fR \fIdesc\fR]
[\fB--skip-suspend\fR]
.PP
vzctl \fBsnapshot-switch\fR | \fBsnapshot-delete\fR \fIctid\fR|\fIname\fR \fB--id\fR \fIuuid\fR
.PP
vzctl \fBsnapshot-mount\fR \fICTID\fR \fB--id\fR \fIuuid\fR \fB--target\fR \fIpath\fR
.PP
vzctl \fBsnapshot-umount\fR \fICTID\fR \fB--id\fR \fIuuid\fR
.PP
vzctl \fBsnapshot-list \fICTID\fR [\fB-H\fR] [\fB-o\fR \fIfield\fR[,\fIfield\fR...] [\fB--id\fR \fIuuid\fR]
.PP
vzctl [\fB--quiet\fR | \fB--verbose\fR] \fBexec\fR | \fBexec2\fR | \fBexec3\fR \fIctid\fR|\fIname\fR \fIcommand\fR [\fIarg\fR ...]
.PP
vzctl \fBconvert\fR \fIctid\fR|\fIname\fR
.PP
vzctl \fBregister\fR \fIpath\fR \fIctid\fR [\fB--force\fR]
.PP
vzctl \fBunregister\fR \fIctid\fR|\fIname\fR
.PP
vzctl \fBcompact\fR \fIctid\fR|\fIname\fR [\fB--defrag\fR] [\fB--dry\fR] [\fB--threshold\fR  \fIvalue\fR] [\fB--delta\fR  \fIvalue\fR]
.PP
vzctl \fB--help\fR | \fB--version\fR
.SH DESCRIPTION
Utility \fBvzctl\fR runs on host server (HS) and performs direct
manipulations with Container (CT).
.PP
Containers can be referred to by numeric \fIctid\fR
(or \fIname\fR which can be assigned with \fB--name\fR option). Note
that CT ID <= 100 are reserved for @PRODUCT_NAME_SHORT@ internal purposes.
.SH OPTIONS
.SS Flags
These flags can be used with any \fBvzctl\fR command.
.IP \fB--quiet\fR 4
Disables logging to a console (a terminal). Logging to a log file
is not affected.
.IP \fB--verbose\fR 4
Sets console (terminal) logging level to maximum value. Logging to a log file
is not affected.
.SS Setting CT parameters
.IP "\fBset\fR \fIctid\fR" 4
This command sets various CT parameters. If flag \fB--save\fR is given,
parameters are saved in CT configuration file \fBve.conf\fR(5).
If CT is currently running, \fBvzctl\fR applies these parameters to CT,
in case \fB--setmode\fR option specified apply logic for parameters that can't
be applied or tunning CT will be changed, if \fBrestart\fR - restart CT,
if \fBignore\fR - ignore error.
Parameters that can not be applied runtime are
\fBquotaugidlimit\fR, \fBbindmount\fR, \fBfeatures\fR.

The following options can be used with \fBset\fR command.
.TP
\fBMiscellaneous parameters\fR
.TP
\fB--setmode\fR \fBrestart\fR|\fBignore\fR
Change default mode for apply parameters that can not be applied on running CT.
\fBrestart\fR - restart CT, \fBignore\fR - ignore error.
.TP
\fB--onboot\fR \fByes\fR|\fBno\fR
Sets whether this CT will be started during system boot up. CT will not be
auto-started during system boot up unless this parameter is set to \fByes\fR.
.TP
\fB--bootorder\fR \fInumber\fR
Sets the boot order priority for this CT. The higher the \fInumber\fR is,
the earlier in the boot process this container starts. By default this
parameter is unset, this is considered to be the lowest priority, so
containers with unset \fBbootorder\fR will start last.
.TP
\fB--root\fR \fIpath\fR
Sets path to root directory for this CT. This is essentially a mount
point for CT root. Argument can contain literal string \fB$VEID\fR, which will
be substituted with numeric CT ID. Changing this parameter is not
recommended, better edit \fBvz\fR(5) global configuration file.
.TP
\fB--userpasswd\fR \fIuser\fR:\fIpassword\fR
Sets password for given user in CT, creating the user if it does not exists.
Note that this option is not saved in configuration file at all (so
\fB--save\fR flag is useless), it is applied to CT (by modifying its
\fB\f(CR/etc/passwd\fR and \fB\f(CR/etc/shadow\fR files).

If the \fB--crypted\fR parameter is specified, the system assumes that the
passwords are encrypted.

In case CT root is not mounted, it is automatically mounted, then all
appropriate file changes are applied, then it is unmounted.

Note that CT area should be created before using this option.
.TP
\fB--disabled\fR \fByes\fR|\fBno\fR
Disable CT start. For force start disabled CT option \fI--force\fR can be used.
.TP
\fB--name\fR \fIname\fR
Bind CT with \fIname\fR, it allows to use name instead of CT ID.
The valid symbols for \fIname\fR are [0-9][aA-Zz][\ -_.] and all international symbols.
.TP
\fB--description\fR \fIdesc\fR
Assign description for CT. It can be viewed by \fBvzlist\fR(8) utility.
.TP
\fB--distribution\fR \fIdist\fR
Assign distribution for CT. It can be viewed by \fBvzlist\fR(8) utility.
.TP
\fBNetwork related parameters\fR
.TP
\fB--ipadd\fR \fIaddr\fR[\fB/\fImask\fR
Adds the IP address and subnet mask for the Container. To assign network masks to Containers operating in the
venet0 mode, the USE_VENET_MASK parameter in the Containers configuration file must be set to
"yes". Note that this option is incremental, so \fIaddr\fR are added to already existing ones.
.TP
\fB--ipdel\fR \fIaddr\fR | \fBall\fR
Removes IP address \fIaddr\fR from CT. If you want to remove all addresses,
use \fB--ipdel all\fR.
.TP
\fB--ext_ipadd\fR \fIaddr\fR
Assign the external IP address to the Container. External IP addresses are considered valid IP addresses by
the venet0 adapter, though they are not set as alias addresses inside Containers and are not announced via
Address Resolution Protocol (ARP). You can assign the same external IP address to several Containers,
irrespective of whether they reside on the same or different Hardware Nodes.
.TP
\fB--ext_ipdel\fR \fIaddr\fR
Remove the external IP address from the Container. To delete all external IP addresses assigned to the
Container, use ext_ipdel all.
.TP
\fB--netif_add\fR \fIifname[\fR,\fImac\fR,\fIhost_ifname\fR,\fIhost_mac]\fR
Adds virtual ethernet device to given CT. Where \fIifname\fR is ethernet
device name in the CT, \fImac\fR its MAC address, \fIhost_ifname\fR is
ethernet device name on the host and \fIhost_mac\fR its MAC address.
MAC addresses has format like XX:XX:XX:XX:XX:XX. All parameters except
\fIifname\fR are option and automatically generated if not specified.

Per-interface configuration.
.br
To select the interface to configure, use \fB--ifname\fR \fIname\fR option.
.br
\fB--mac\fR \fIXX:XX:XX:XX:XX:XX\fR - MAC address of interface inside CT
.br
\fB--host_ifname\fR \fIname\fR - interface name for virtual interface on host server
.br
\fB--host_mac\fR \fIXX:XX:XX:XX:XX:XX\fR - MAC address of interface on host server
.br
\fB--gw\fR \fIipaddr\fR - default IPv4 gateway for interface
.br
\fB--gw6\fR \fIipaddr\fR - default IPv6 gateway for interface
.br
\fB--ipadd\fR \fIipaddr\fR - add IP address(es) to interface
.br
\fB--ipdel\fR \fIipaddr\fR - delete IP address(es) from interface
.br
\fB--dhcp\fR \fByes\fR|\fBno\fR - turn on/off IPv4 dhcp
.br
\fB--dhcp6\fR \fByes\fR|\fBno\fR - turn on/off IPv6 dhcp
.br
\fB--configure\fR \fBnone\fR|\fBall\fR - apply/ignore the network settings (gw,ip,dhcp)
from the Container configuration file.
Configuring any of the network settings automatically sets this option to "all".
.br
\fB--network\fR \fIid\fR - connect virtual interface to virtual network
with the given \fIid\fR.
The valid symbols for \fInetwork\fR are [0-9][aA-Zz][\ -_.#()] and all
international symbols.
.br
\fB--mac_filter\fR \fBon\fR|\fBoff\fR - enable/disable packets filtering by MAC
address and MAC changing on veth device inside CT.
.TP
\fB--netif_del\fR \fIdev_name\fR
Removes virtual ethernet device from CT.
.TP
\fB--hostname\fR \fIname\fR
Sets CT hostname and writes it to the appropriate file inside CT
(distribution-dependent).
.TP
\fB--nameserver\fR \fIaddr\fR
Sets DNS server IP address for CT. If you want to set several nameservers,
you should do it at once, so use \fB--nameserver\fR option multiple times
in one call to \fBvzctl\fR, as all the name server values set in previous
calls to \fBvzctl\fR gets overwritten.
.TP
\fB--searchdomain\fR \fIname\fR
Sets DNS search domains for CT. If you want to set several search domains,
you should do it at once, so use \fB--searchdomain\fR option multiple times
in one call to \fBvzctl\fR, as all the search domain values set in previous
calls to \fBvzctl\fR gets overwritten.
.TP
.TP
\fBResource limits\fR
The following options sets barrier and limit for various user beancounters.
Each option requires one or two arguments. In case of one argument,
\fBvzctl\fR sets barrier and limit to the same value. In case of
two colon-separated arguments, the first is a barrier,
and the second is a limit.

Arguments are in items, pages or bytes. Note that the page size
is architecture-specific, it is 4096 bytes on IA32 platform.

There is an ability to accept different suffixes for
set parameters (except the parameters that name started with \fBnum\fR).
E.g. \fBvzctl set XXX --privvmpages 5M:6M\fR should set \fBprivvmpages\fR
barrier to 5 megabytes and limit to 6 megabytes.

Available suffixes are:
.br
\fBt\fR, \fBT\fR -- terabytes
.br
\fBg\fR, \fBG\fR -- gigabytes
.br
\fBm\fR, \fBM\fR -- megabytes
.br
\fBk\fR, \fBK\fR -- kilobytes
.br
\fBp\fR, \fBP\fR -- pages (page is 4096 bytes on x86 architecture, other
arches may differ)
.TP
\fB-p\fR, \fB--numproc\fR \fIitems\fR[:\fIitems\fR]
Maximum number of processes and kernel-level threads.
Setting the barrier and
the limit to different values does not make practical sense.
.TP
\fB--vmguarpages\fR \fIpages\fR[:\fIpages\fR]
Memory allocation guarantee. This parameter controls how much memory is
available to CT. The barrier is the amount
of memory that CT's applications are guaranteed to be able to allocate.
The meaning of the limit is currently unspecified; it should be set to
2,147,483,647.
.TP
\fB-l\fR, \fB--lockedpages\fR \fIpages\fR[:\fIpages\fR]
Maximum number of pages acquired by \fBmlock\fR(2).
.TP
\fB--privvmpages\fR \fIpages\fR[:\fIpages\fR]
Allows controlling the amount of memory allocated by the applications.
For shared (mapped as \fBMAP_SHARED\fR) pages, each CT really using a memory
page is charged for the fraction of the page (depending on the number of
others using it). For "potentially private" pages (mapped as
\fBMAP_PRIVATE\fR), CT is charged either for a fraction of the size or for
the full size if the allocated address space. It the latter case, the physical
pages associated with the allocated address space may be in memory, in swap
or not physically allocated yet.

The barrier and the limit of this parameter
control the upper boundary of the total size of allocated memory. Note that
this upper boundary does not guarantee that CT will be able to allocate that
much memory. The primary mechanism to control memory allocation is
the \fB--vmguarpages\fR guarantee.
.TP
\fB--shmpages\fR \fIpages\fR[:\fIpages\fR]
Maximum IPC SHM segment size.
Setting the barrier and
the limit to different values does not make practical sense.
.TP
\fB-n\fR, \fB--numfile\fR \fIitems\fR[:\fIitems\fR]
Maximum number of open files.
Setting the barrier and
the limit to different values does not make practical sense.
.TP
\fB-f\fR, \fB--numflock\fR \fIitems\fR[:\fIitems\fR]
Maximum number of file locks. Safety gap should be between barrier and limit.
.TP
\fB-t\fR, \fB--numpty\fR \fIitems\fR[:\fIitems\fR]
Number of pseudo-terminals (PTY). Note that in @PRODUCT_NAME_SHORT@ each CT
can have no more than 255 PTYs. Setting the barrier and
the limit to different values does not make practical sense.
.TP
\fB-i\fR, \fB--numsiginfo\fR \fIitems\fR[:\fIitems\fR]
Number of siginfo structures.
Setting the barrier and
the limit to different values does not make practical sense.
.TP
\fB-e\fR, \fB--numiptent\fR \fInum\fR[:\fInum\fR]
Number of iptables (netfilter) entries.
Setting the barrier and
the limit to different values does not make practical sense.
.TP
\fB--physpages\fR \fIpages\fR[:\fIpages\fR]
This parameter limits the physical memory (RAM) available to processes inside a container.
The barrier is ignored, and the limit sets the limit.
.TP
\fB--swappages\fR \fIpages\fR[:\fIpages\fR]
This parameter limits the amount of swap space that can be allocated to processes running in a container.
Container processes can use swap even if the host has available memory. Use of swap (as opposed to only reserving it)
significantly slows down both the container and the node.
.TP
\fB--ram\fR \fIbytes\fR
The amount of RAM that can be used by the processes of a Container, in bytes. You can use the following suffixes to set RAM in other measurement units:
.br
\fBk\fR, \fBK\fR -- kilobytes
.br
\fBm\fR, \fBM\fR -- megabytes
.br
\fBg\fR, \fBG\fR -- gigabytes
.br
\fBt\fR, \fBT\fR -- terabytes
.TP
\fB--swap\fR \fIbytes\fR
The amount of swap space that can be used by the Container for swapping out memory once the RAM is exceeded, in bytes.
Container processes can use swap even if the host has available memory. Use of swap (as opposed to only reserving it)
significantly slows down both the container and the node. You can use the following suffixes to set swap in other measurement units:
.br
\fBk\fR, \fBK\fR -- kilobytes
.br
\fBm\fR, \fBM\fR -- megabytes
.br
\fBg\fR, \fBG\fR -- gigabytes
.br
\fBt\fR, \fBT\fR -- terabytes
.TP
\fB--vm_overcommit\fR \fIN\fR
This parameter controls the memory allocation guarantee. It is calculated as
(physpages + swappages) * N. By default, the parameter is equal to 1.
.TP
\fB--memguarantee\fR \fIauto|value\fR
Set the amount of memory (RAM) that will be guaranteed to a Container.
The guaranteed memory is a percentage of total RAM that is set for the
Container with the '--memsize' option. By default, memory guarantee is
set to 'auto' (0% - no gurantee).
.TP
.TP
\fB-e\fR, \fB--numnetif\fR \fInum\fR
Maximum number of network interfaces.
\fB--pagecache-isolation\fR \fIyes|no\fR
By default page cache evicted from a memory cgroup (Container) on local
pressure is stored in Nodes memory unmodified in case there is free RAM on
the Node, thus the pages can be fetched back later without costly disk
accesses improving the overall Node performance.
Enable page cache isolation (set to \fIyes\fR) tuned off this logic.
.TP
\fBCPU fair scheduler parameters\fR
These parameters control CPU usage by CT.
.TP
\fB--cpuunits\fR \fInum\fR
sets CPU weight for CT. Argument is positive number, which passed to
and used in kernel fair scheduler. The larger the number is, the more CPU time
this CT get. Maximum value is 500000, minimal is 8. Number is relative to
weights of all the other running CTs. If cpuunits not specified the default value
1000 is used.
.TP
\fB--cpulimit\fR \fInum\fR
Sets the CPU limit, in percent or megahertz (MHz), the Container is not allowed to exceed.
By default, the limit is set in percent. To specify the limit in MHz, specify "m" after the value.
Note: If the computer has 2 CPUs, the total CPU time equals 200%.
.TP
\fB--cpus\fR \fInum\fR
sets number of CPUs available in the CT.
.TP
\fB--cpumask\fR {\fIN[,N][N-N]\fR|\fBall\fR|\fBauto\fR}

etc/bash_completion.d/vzctl.sh
#!/bin/bash
# Copyright (c) 2016-2019 Virtuozzo International GmbH.
# All rights reserved.

_get_ves()
{
	local cmd=$1
	local rm_empty='sed s/^-[[:space:]]*//'
	[ "$EUID" -ne 0 ] && exit 1
	case $cmd in
		create)
			# create a new VEID
			echo "`uuidgen`"
			;;
		start|mount|umount|destroy|delete)
			# stopped VEs
			/usr/sbin/vzlist -H -S -octid
			/usr/sbin/vzlist -H -S -oname | $rm_empty
			;;
		stop|enter|exec*|execaction|suspend)
			# running VEs
			/usr/sbin/vzlist -H -octid
			/usr/sbin/vzlist -H -oname | $rm_emtpy
			;;
		*)
			# All VEs
			/usr/sbin/vzlist -H -a -octid
			/usr/sbin/vzlist -H -a -oname | $rm_empty
			;;
	esac
}

_vzctl()
{
	#echo "ARGS: $*"
	#echo "COMP_WORDS: $COMP_WORDS"
	#echo "COMP_CWORD: $COMP_CWORD"
	#echo "COMP_WORDS[1]: ${COMP_WORDS[1]}"

	COMPREPLY=()
	local cur=${COMP_WORDS[COMP_CWORD]}
	local prev=${COMP_WORDS[COMP_CWORD-1]}

	local vzctl_common_opts="--quiet --verbose --help --version"
	local vzctl_cmds="create destroy delete mount umount chkpnt suspend restore \
		set unset start stop restart status enter exec exec2 exec3 runscript \
        setrate pause resume monitor register unregister execaction \
		compact console convert snapshot snapshot-switch snapshot-delete \
		snapshot-mount snapshot-umount snapshot-list"
	local vzctl_create_opts="--ostemplate --config --private --root \
		--ipadd --hostname --name --description --skip_app_templates \
		--encryption-keyid"
	local vzctl_compact_opts="--defrag --dry --threshold --delta"
	local vzctl_set_opts="--save --onboot \
		--numproc --oomguarpages \
		--lockedpages --privvmpages --shmpages \
		--numfile --numflock --numpty --numsiginfo \
		--numiptent --physpages --swappages --diskspace \
		--quotaugidlimit --cpuunits --cpulimit \
		--cpus --ipadd --ipdel \
		--root --hostname --nameserver --searchdomain \
		--userpasswd --onboot --rate --ratebound \
		--noatime --bindmount_add --bindmount_del \
		--capability --devices  --devnodes \
		--netdev_add --netdev_del \
		--netif_add --netif_del \
		--disabled --applyconfig --setmode \
		--bootorder --ha_enable --ha_prio --encryption-keyid \
		--reencrypt --wipe"
	local vzctl_snapshot_create_opts="--id --name --description"
	local vzctl_snapshot_mount_opts="--id --target"
	local vzctl_snapshot_list_opts="-H --no-header -o --output --id \
		-L --list --help"
	local vzctl_console_opts="-s --start 1 2"


	local cap_names="chown dac_override dac_read_search fowner fsetid kill
		setgid setuid setpcap linux_immutable net_bind_service
		net_broadcast net_admin net_raw ipc_lock ipc_owner
		sys_module sys_rawio sys_chroot sys_ptrace sys_pacct
		sys_admin sys_boot sys_nice sys_resource sys_time
		sys_tty_config mknod lease setveid ve_admin"



	case $COMP_CWORD in
	1)
		# command or global option
		COMPREPLY=( $( compgen -W "$vzctl_cmds $vzctl_common_opts" -- $cur ) )
		;;

	2)
		case "$prev" in
		--help|--version)
			COMPREPLY=()
			;;
		--*)
			# command
			COMPREPLY=( $( compgen -W "$vzctl_cmds" -- $cur ) )
			;;
		*)
			# VEID
			COMPREPLY=( $( compgen -W "$(_get_ves $prev)" -- $cur ) )
			;;
		esac
		;;

	*) # COMP_CWORD >= 3
		if [[ $COMP_CWORD -eq 3 && ! $prev != '^[1-9][0-9]*$' ]]; then
			# VEID
			COMPREPLY=( $( compgen -W "$(_get_ves $prev)" -- $cur ) )
		else

			# flag or option
			case $prev in
			--ostemplate)
				COMPREPLY=( $( compgen -W "$(/usr/sbin/vzpkg list -O -q;
					/usr/sbin/vzpkg list -O -q --available)" -- $cur ) )
				;;
			--onboot|--disabled|--noatime|--ha_enable)
				COMPREPLY=( $( compgen -W "yes no" -- $cur ) )
				;;
			--setmode)
				COMPREPLY=( $( compgen -W "restart ignore" -- $cur ) )
				;;
			--config|--applyconfig)
				local configs=$(command ls -1 /etc/vz/conf/*.conf-sample | \
						    cut -d "-" -f 2- | sed -e 's#.conf-sample$##')

				configs=${configs/.conf-sample/}
				COMPREPLY=( $( compgen -W "$configs" -- $cur ) )
				;;
			--netdev*)
				local devs=`ip addr show | awk '/^[0-9]/ && /UP/ && !/venet/ && !/lo/ \
						    { print $2 }' | sed s/://`

				COMPREPLY=( $( compgen -W "$devs" -- $cur ) )
				;;
			--meminfo)
				local meminfo="none privvmpages:1"
				COMPREPLY=( $( compgen -W "$meminfo" -- $cur ) )
				;;
			--ioprio)
				# ioprio range: 0 - 7 (default : 4)
				local ioprio='0 1 2 3 4 5 6 7'
				COMPREPLY=( $( compgen -W "$ioprio" -- $cur ) )
				;;
			--capability)
				# capname:on|off
				COMPREPLY=( $( compgen -W "$cap_names" -- $cur ) )
				# FIXME: add :on or :off -- doesn't work :(
#				if [[ ${#COMPREPLY[@]} -le 1 ]]; then
#					if [[ $cur =~ ":" ]]; then
#						cap=${cur%%:*}
#					else
#						cap=${COMPREPLY[0]%%:*}
#					fi
					# Single match: add :on|:off
#					COMPREPLY=( $(compgen -W \
#					"${cap}:on ${cap}:off" -- $cur) )
#				fi
				;;
			--devnodes)
				# FIXME: device:r|w|rw|none
				local devs=''
				COMPREPLY=( $( compgen -W "$devs" -- $cur ) )
				;;
			--devices)
				# FIXME: b|c:major:minor|all:r|w|rw
				local devices=''
				COMPREPLY=( $( compgen -W "$devices" -- $cur ) )
				;;
			--ipdel)
				[ "$EUID" -ne 0 ] && exit 1
				# Get VEID
				local ve=${COMP_WORDS[2]}
				if [[ ! ${ve} != '^[1-9][0-9]*$' ]] ; then
					# --verbose or --quiet used
					ve=${COMP_WORDS[3]}
				fi
				# VENAME or VEID ?
				LIST_OPT=`echo $ve | awk '/[a-zA-Z]/ {print "-N"}'`
				local ips="`/usr/sbin/vzlist -H -o ip $LIST_OPT $ve | grep -vi -`"
				COMPREPLY=( $( compgen -W "$ips all" -- $cur ) )
				;;
			--netif_del)
				local netif
				if [ -f /proc/vz/veth ]; then
					local ve=${COMP_WORDS[2]}
					if [[ ! ${ve} != '^[1-9][0-9]*$' ]] ; then
						# --verbose or --quiet used
						ve=${COMP_WORDS[3]}
					fi
					# get veth(ernet) device of VE, proc and config
					local netif_proc="`awk "/${ve}/ { print \\$4 }" /proc/vz/veth`"
					local netif_conf="`grep ^NETIF /etc/vz/conf/${ve}.conf | \
						sed -e 's/"/\n/g' -e 's/;/\n/g' -e 's/,/\n/g' -e 's/=/ /g' | awk '/^ifname/ {print \$2}'`"
					netif="$netif_proc $netif_conf"
				fi
				COMPREPLY=( $( compgen -W "$netif" -- $cur ) )
				;;
			--private|--root)
				# FIXME
				# Dir autocompletion works bad since there is
				# a space added. Alternatively, we could use
				# -o dirname option to 'complete' -- but it
				# makes no sense for other parameters (UBCs
				# etc). So no action for now.
				;;
			-s|--start)
				# vzctl console -s|--start
				# Complete tty number, except 1 and 2
				# which already should have getty running
				COMPREPLY=( $( compgen -W "3 4 5 6 7 8 9" -- $cur ) )
				;;
			--id|--uuid)
				# Only allow this completion
				# when user has root privileges
				[ "$EUID" -ne 0 ] && exit 1
				# Get command and CTID
				local cmd=${COMP_WORDS[1]}
				local ve=${COMP_WORDS[2]}
				if [ ${cmd:0:2} = "--" ] ; then
					# --verbose or --quiet used
					cmd=${COMP_WORDS[2]}
					ve=${COMP_WORDS[3]}
				fi
				local ids
				case ${cmd} in
				    snapshot-umount)
					# For umount, only give mounted IDs
					ids="`/usr/sbin/vzctl snapshot-list $ve -H -ouuid,device 2>/dev/null | awk 'NF==2 {print $1}' | tr -d '{}'`"
					;;
				    snapshot-*)
					# For other commands, give all IDs
					ids="`/usr/sbin/vzctl snapshot-list $ve -H -ouuid 2>/dev/null | tr -d '{}'`"
					;;
				    *)
					# Do not give anything for other cmds
					# including 'snapshot' itself
					# since it requires a new unused ID
					exit 0
					;;
				esac
				COMPREPLY=( $( compgen -W "$ids" -- $cur ) )
				;;
			*)
				if [[ "${prev:0:2}" != "--" || "$prev" = "--save" || "$prev" = "--dry" || "$prev" = "--defrag" ]]; then
					# List options
					local cmd=${COMP_WORDS[1]}
					if [ ${cmd:0:2} = "--" ] ; then
						# --verbose or --quiet used
						cmd=${COMP_WORDS[2]}
					fi

					case "$cmd" in
					create)
						COMPREPLY=( $( compgen -W "$vzctl_create_opts" -- $cur ) )
						;;
					set)
						COMPREPLY=( $( compgen -W "$vzctl_set_opts" -- $cur ) )
						;;
					chkpnt|restore)
						COMPREPLY=( $( compgen -W "--dumpfile" -- $cur ) )
						;;
					start)
						COMPREPLY=( $( compgen -W "--wait" -- $cur ) )
						;;

					stop)
						COMPREPLY=( $( compgen -W "--fast --skip-umount" -- $cur ) )
						;;
					snapshot)
						COMPREPLY=( $( compgen -W "$vzctl_snapshot_create_opts" -- $cur ) )
						;;
					snapshot-switch|snapshot-delete|snapshot-umount)
						COMPREPLY=( $( compgen -W "--id" -- $cur ) )
						;;
					snapshot-mount)
						COMPREPLY=( $( compgen -W "$vzctl_snapshot_mount_opts" -- $cur ) )
						;;
					snapshot-list)
						COMPREPLY=( $( compgen -W "$vzctl_snapshot_list_opts" -- $cur ) )
						;;
					console)
						COMPREPLY=( $( compgen -W "$vzctl_console_opts" -- $cur ) )
						;;
					compact)
						COMPREPLY=( $( compgen -W "$vzctl_compact_opts" -- $cur ) )
						;;

					*)
						;;
					esac
				else
					# Option that requires an argument
					# which we can't autocomplete
					COMPREPLY=( $( compgen -W "" -- $cur ) )
				fi
				;;
			esac
		fi
	esac

	return 0
}

complete -F _vzctl vzctl

# EOF
